jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java

```diff
@@ -25,192 +25,256 @@
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.List;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
 import javax.annotation.Nullable;
 
 /**
  * Runs steps for building an image.
  *
- * <p>Use by constructing the runner and calling {@code run...} each step. Make sure that steps are
- * run before other steps that depend on them. Wait on the last step.
+ * <p>Use by first calling {@link #begin} and then calling the individual step running methods. Note
+ * that order matters, so make sure that steps are run before other steps that depend on them. Wait
+ * on the last step by calling the respective {@code wait...} methods.
  */
 public class StepsRunner {
 
-  private final ListeningExecutorService listeningExecutorService;
-  private final BuildConfiguration buildConfiguration;
+  /** Holds the individual steps. */
+  private static class Steps {
 
-  @Nullable private RetrieveRegistryCredentialsStep retrieveTargetRegistryCredentialsStep;
-  @Nullable private AuthenticatePushStep authenticatePushStep;
-  @Nullable private PullBaseImageStep pullBaseImageStep;
-  @Nullable private PullAndCacheBaseImageLayersStep pullAndCacheBaseImageLayersStep;
+    @Nullable private RetrieveRegistryCredentialsStep retrieveTargetRegistryCredentialsStep;
+    @Nullable private AuthenticatePushStep authenticatePushStep;
+    @Nullable private PullBaseImageStep pullBaseImageStep;
+    @Nullable private PullAndCacheBaseImageLayersStep pullAndCacheBaseImageLayersStep;
 
-  @Nullable
-  private ImmutableList<BuildAndCacheApplicationLayerStep> buildAndCacheApplicationLayerSteps;
+    @Nullable
+    private ImmutableList<BuildAndCacheApplicationLayerStep> buildAndCacheApplicationLayerSteps;
 
-  @Nullable private PushLayersStep pushBaseImageLayersStep;
-  @Nullable private PushLayersStep pushApplicationLayersStep;
-  @Nullable private BuildImageStep buildImageStep;
-  @Nullable private PushContainerConfigurationStep pushContainerConfigurationStep;
-  @Nullable private PushImageStep pushImageStep;
-  @Nullable private LoadDockerStep loadDockerStep;
-  @Nullable private WriteTarFileStep writeTarFileStep;
-
-  public StepsRunner(BuildConfiguration buildConfiguration) {
-    this.buildConfiguration = buildConfiguration;
+    @Nullable private PushLayersStep pushBaseImageLayersStep;
+    @Nullable private PushLayersStep pushApplicationLayersStep;
+    @Nullable private BuildImageStep buildImageStep;
+    @Nullable private PushContainerConfigurationStep pushContainerConfigurationStep;
+    @Nullable private PushImageStep pushImageStep;
+    @Nullable private LoadDockerStep loadDockerStep;
+    @Nullable private WriteTarFileStep writeTarFileStep;
+  }
 
+  /**
+   * Starts building the steps to run.
+   *
+   * @param buildConfiguration the {@link BuildConfiguration}
+   * @return a new {@link StepsRunner}
+   */
+  public static StepsRunner begin(BuildConfiguration buildConfiguration) {
```

Comment:
As a mere static instance creator, wouldn't below make more sense?
```java
public static StepsRunner begin(BuildConfiguration buildConfiguration) {
  return new StepsRunner(buildConfiguration);
}
```
(While just making the `StepsRunner` constructor `private` from `public`, without really touching it.) Not getting why taking out part of the existing, internal constructor code into the static initializer.

Reply:
This is mostly following the idea that potentially costly logic should go in a static initializer rather than a constructor, which should not have any costly logic. This might not be necessary here but it's under the assumption that creating an `ExecutorService` could potentially be a costly operation.